salt/transport/rabbitmq.py

```diff
@@ -0,0 +1,1243 @@
+            log.trace(
+                "Inserted key into loop_instance_map id %s for key %s and process %s",
+                id(loop_instance_map),
+                key,
+                os.getpid(),
+            )
+        else:
+            with obj._refcount_lock:
+                obj._refcount += 1
+            log.debug("Re-using AsyncRabbitMQReqChannel for %s", key)
+        return obj
+
+    def __deepcopy__(self, memo):
+        cls = self.__class__
+        # pylint: disable=too-many-function-args
+        result = cls.__new__(cls, copy.deepcopy(self.opts, memo))
+        # pylint: enable=too-many-function-args
+        memo[id(self)] = result
+        for key in self.__dict__:
+            if key in ("_io_loop", "_refcount", "_refcount_lock"):
+                continue
+                # The _io_loop has a thread Lock which will fail to be deep
+                # copied. Skip it because it will just be recreated on the
+                # new copy.
+            if key == "message_client":
+                # Recreate the message client because it will fail to be deep
+                # copied. The reason is the same as the io_loop skip above.
+                setattr(
+                    result,
+                    key,
+                    AsyncReqMessageClientPool(
+                        result.opts,
+                        args=(result.opts, self.master_uri,),
+                        kwargs={"io_loop": self._io_loop},
+                    ),
+                )
+
+                continue
+            setattr(result, key, copy.deepcopy(self.__dict__[key], memo))
+        return result
+
+    @classmethod
+    def force_close_all_instances(cls):
+        """
+        Will force close all instances
+
+        TODO: PR - add a justification why we need this for RMQ. Do we need this?
+
+        :return: None
+        """
+        for weak_dict in list(cls.instance_map.values()):
+            for instance in list(weak_dict.values()):
+                instance.close()
+
+    @classmethod
+    def __key(cls, opts, **kwargs):
+        return (
+            opts["pki_dir"],  # where the keys are stored
+            opts["id"],  # minion ID
+            kwargs.get("master_uri", opts.get("master_uri")),  # master ID
+            kwargs.get("crypt", "aes"),  # TODO: use the same channel for crypt
+        )
+
+    # has to remain empty for singletons, since __init__ will *always* be called
+    def __init__(self, opts, **kwargs):
+        pass
+
+    # an init for the singleton instance to call
+    def __singleton_init__(self, opts, **kwargs):
+        self.opts = dict(opts)
+        self.ttype = "rabbitmq"
+
+        # crypt defaults to 'aes'
+        self.crypt = kwargs.get("crypt", "aes")
+
+        if "master_uri" in kwargs:
+            self.opts["master_uri"] = kwargs["master_uri"]
+
+        self._io_loop = kwargs.get("io_loop")
+        if self._io_loop is None:
+            self._io_loop = tornado.ioloop.IOLoop.current()
```

Comment:
**suggestion** `or` suggestion applies here as well.

Reply:
Sure. Note that some of this implementation came from zeromq.py. One of the next steps is to refactor common transport code into a base class.